### ISSUE TITLE
Can't make private JDK fields accessible in Java 9

### ISSUE THREAD

**Issue by chrisseaton** [jnr/jnr-ffi#105]
https://github.com/jnr/jnr-ffi/blob/1deb47d686ae384ad9b87fadbcc42aa0cce6da56/src/main/java/jnr/ffi/util/ref/internal/Finalizer.java#L203

This code seems to have been imported from Guava. I've mentioned the issue there as well.

https://github.com/google/guava/issues/2571#issuecomment-273750815

**Comment by headius** [comment]
Ahh, I think I've seen this warning. The setAccessible uses throughout Java-land are going to require some digging into modules.

**Comment by chrisseaton** [comment]
But is it a problem or not? Either we need it or we don't surely?

**Comment by headius** [comment]
The warning is telling us that the finalizer threads used to clean up native references are going to inherit thread locals from some parent thread. I can see this could be a problem if those references are not intended to be kept alive.

However, it appears that Java 9 provides a new Thread constructor that takes a boolean for "inheritThreadLocals" so this should be a simple thing to fix.

**Comment by headius** [comment]
@eamonnmcmanus ^

**Comment by eamonnmcmanus** [comment]
Thanks! Yes, that's basically what I had in mind. The most recent Guava version already has Finalizer being a Runnable rather than a Thread, so the change will be a bit simpler. On the other hand, there doesn't seem to be any test coverage for the problem that this InheritableThreadLocal code is trying to resolve, so we'll probably want to add that to [FinalizableReferenceQueueClassLoaderUnloadingTest](https://github.com/google/guava/blob/master/guava-tests/test/com/google/common/base/FinalizableReferenceQueueClassLoaderUnloadingTest.java).

**Comment by headius** [comment]
We can't depend on something as large as Guava, so I hope it's ok that someone (else) copied this code in :-)

**Comment by eregon** [comment]
This now triggers a warning on Java 9 from https://github.com/jnr/jnr-ffi/blob/469727224e3a73f2db398df1113921f0e9d6313c/src/main/java/jnr/ffi/util/ref/internal/Finalizer.java#L101
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by jnr.ffi.util.ref.internal.Finalizer (file:/home/eregon/code/truffleruby-ws/truffleruby/mxbuild/dists/truffleruby.jar) to field java.lang.Thread.inheritableThreadLocals
WARNING: Please consider reporting this to the maintainers of jnr.ffi.util.ref.internal.Finalizer
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

Probably just trying the Java 9 constructor first and then not trying to access the field would avoid this warning.

**Comment by eregon** [comment]
Fix for the above warning in https://github.com/jnr/jnr-ffi/pull/143

**Comment by headius** [comment]
Additional fixes from #143 released in 2.1.7.
